File: packages/saa-proactive-js/src/trigger-server.js
// @ts-check
// SPDX-License-Identifier: Apache-2.0
/**
 * TriggerHub is an in-process pub/sub for proactive-turn events. It
 * exists so framework overlays can stop hand-rolling the
 * back-end-webhook → SSE-fan-out → browser path: every proactive
 * trigger ends with a back-end HTTP POST, a fan-out to one-or-more
 * connected browsers (or other subscribers), and a JSON payload that
 * carries instructions for the agent's opening turn. The shape is the
 * same across OpenAI Realtime and ElevenLabs CAI browser overlays; the
 * hub is the shared piece.
 *
 * The hub deliberately does NOT bind to a specific HTTP framework
 * (Express / Fastify / Hono / FastAPI). It exposes:
 *
 *   - publish(event)        — call from your HTTP POST handler.
 *   - subscribe()           — returns an async iterator of events.
 *   - sseStream(subscriber) — convenience: returns a ReadableStream
 *                             ready to hand to a fetch-Response or
 *                             Node ServerResponse for SSE.
 *
 * @see ../README.md for an integration recipe.
 */

/**
 * @typedef {{
 *   push(event: TriggerEvent): void,
 *   events(): AsyncGenerator<TriggerEvent, void, void>,
 *   close(): void,
 * }} Subscriber
 */

/**
 * @typedef {{ instructions: string, [key: string]: unknown }} TriggerEvent
 */

export class TriggerHub {
  constructor() {
    /** @type {Set<Subscriber>} */
    this._subscribers = new Set();
  }

  /** @returns {number} */
  get subscriberCount() {
    return this._subscribers.size;
  }

  /**
   * @param {TriggerEvent} event
   * @returns {number} the number of subscribers that received the event
   */
  publish(event) {
    if (!event || typeof event.instructions !== "string" || event.instructions.length === 0) {
      throw new TypeError(
        "[saa-proactive] TriggerHub.publish(event): event must have a non-empty 'instructions' string"
      );
    }
    let fanout = 0;
    for (const sub of this._subscribers) {
      try {
        sub.push(event);
        fanout++;
      } catch {
        // a subscriber whose queue is full / aborted is dropped silently;
        // it will be GCed when its stream ends.
      }
    }
    return fanout;
  }

  /** @returns {Subscriber} */
  subscribe() {
    /** @type {TriggerEvent[]} */
    const queue = [];
    /** @type {((value: TriggerEvent | null) => void) | null} */
    let resolveNext = null;
    let closed = false;
    // Matches the Python twin (asyncio.Queue(maxsize=64)). A slow
    // subscriber drops events on overflow rather than growing memory
    // without bound.
    const QUEUE_LIMIT = 64;

    const subscriber = /** @type {Subscriber} */ ({
      push(event) {
        if (closed) return;
        if (resolveNext) {
          const r = resolveNext;
          resolveNext = null;
          r(event);
        } else if (queue.length < QUEUE_LIMIT) {
          queue.push(event);
        }
        // else: queue saturated; drop on overflow.
      },
      events: async function* () {
        try {
          while (!closed) {
            let event;
            if (queue.length > 0) {
              event = queue.shift();
            } else {
              event = await new Promise((resolve) => { resolveNext = resolve; });
            }
            // Check closed AFTER the await: close() wakes the awaiting
            // promise (with null) to unblock the generator; we must NOT
            // yield that wake-up signal to the consumer.
            if (closed) return;
            yield event;
          }
        } finally {
          subscriber.close();
        }
      },
      close: () => {
        if (closed) return;
        closed = true;
        this._subscribers.delete(subscriber);
        if (resolveNext) {
          // Wake the awaiting next() so the generator can exit. The
          // events() loop checks `closed` and returns without yielding,
          // so this null never reaches consumers.
          const r = resolveNext;
          resolveNext = null;
          try { r(null); } catch {}
        }
      },
    });

    this._subscribers.add(subscriber);
    return subscriber;
  }

  /**
   * Return a ReadableStream that emits `event: trigger` SSE frames for
   * a subscriber. Handy for `new Response(stream, {headers: {"content-type": "text/event-stream"}})`
   * in a Fetch-style HTTP server (Hono / Bun / Cloudflare Workers) or
   * for `stream.pipe(res)` in Node 18+ http servers.
   *
   * @param {Subscriber} [subscriber] if omitted, a fresh subscriber is created
   * @returns {ReadableStream<Uint8Array>}
   */
  sseStream(subscriber) {
    const sub = subscriber ?? this.subscribe();
    const encoder = new TextEncoder();
    return new ReadableStream({
      async start(controller) {
        // Comment line on connect so reverse proxies don't time out.
        controller.enqueue(encoder.encode(": connected\n\n"));
        try {
          for await (const event of sub.events()) {
            controller.enqueue(
              encoder.encode(`event: trigger\ndata: ${JSON.stringify(event)}\n\n`)
            );
          }
        } catch (err) {
          // close the stream cleanly on any iteration error
        } finally {
          try { controller.close(); } catch {}
        }
      },
      cancel() {
        sub.close();
      },
    });
  }
}

File: packages/saa-proactive-py/src/saa_proactive/trigger_server.py
"""TriggerHub — in-process pub/sub for proactive-turn events.

Used by framework overlays to relay back-end ``POST /trigger`` HTTP
webhooks to one or more connected browsers via Server-Sent Events.
Every proactive trigger ends with a back-end HTTP POST, a fan-out to
one-or-more connected browsers (or other subscribers), and a JSON
payload that carries instructions for the agent's opening turn. The
hub is the shared piece.

The hub deliberately does NOT bind to a specific HTTP framework. It
exposes:

* :meth:`TriggerHub.publish` — call from your HTTP POST handler.
* :meth:`TriggerHub.subscribe` — returns an async iterator of events.
* :meth:`TriggerHub.fastapi_router` — optional FastAPI integration
  (only available if FastAPI is installed; ``pip install
  attenlabs-saa-proactive[fastapi]``).
"""
from __future__ import annotations

import asyncio
import json
from dataclasses import dataclass
from typing import Any, AsyncIterator, Dict, Set


@dataclass
class TriggerEvent:
    """A single proactive-turn trigger payload."""

    instructions: str
    extra: Dict[str, Any]

    def to_json(self) -> str:
        payload: Dict[str, Any] = {"instructions": self.instructions}
        payload.update(self.extra)
        return json.dumps(payload, ensure_ascii=False)


class _Subscriber:
    """One subscriber. Holds an internal queue and a close flag."""

    def __init__(self, parent: "TriggerHub") -> None:
        self._parent = parent
        self._queue: asyncio.Queue[TriggerEvent] = asyncio.Queue(maxsize=64)
        self._closed = False

    def push(self, event: TriggerEvent) -> None:
        if self._closed:
            return
        try:
            self._queue.put_nowait(event)
        except asyncio.QueueFull:
            # Drop on overflow; the subscriber is too slow.
            pass

    async def events(self) -> AsyncIterator[TriggerEvent]:
        try:
            while not self._closed:
                event = await self._queue.get()
                if self._closed:
                    return
                yield event
        finally:
            self.close()

    def close(self) -> None:
        if self._closed:
            return
        self._closed = True
        self._parent._unregister(self)

    async def sse_lines(self) -> AsyncIterator[bytes]:
        """Yield Server-Sent Events frames for this subscriber.

        First yields ``: connected\\n\\n`` so reverse proxies don't time
        out, then ``event: trigger\\ndata: {...}\\n\\n`` per event. The
        generator exits when the subscriber is closed (e.g. the HTTP
        client disconnects and the server cancels the response stream).
        """
        yield b": connected\n\n"
        try:
            async for event in self.events():
                yield f"event: trigger\ndata: {event.to_json()}\n\n".encode("utf-8")
        except asyncio.CancelledError:
            return


class TriggerHub:
    """In-process pub/sub for proactive-turn events."""

    def __init__(self) -> None:
        self._subscribers: Set[_Subscriber] = set()

    @property
    def subscriber_count(self) -> int:
        return len(self._subscribers)

    def publish(self, instructions: str, **extra: Any) -> int:
        """Fan out a trigger to every connected subscriber.

        :returns: the number of subscribers that received the event.
        :raises ValueError: if ``instructions`` is empty or not a string.
        """
        if not isinstance(instructions, str) or not instructions:
            raise ValueError(
                "[saa-proactive] TriggerHub.publish requires a non-empty 'instructions' string"
            )
        event = TriggerEvent(instructions=instructions, extra=dict(extra))
        fanout = 0
        for sub in list(self._subscribers):
            try:
                sub.push(event)
                fanout += 1
            except Exception:
                pass
        return fanout

    def subscribe(self) -> _Subscriber:
        sub = _Subscriber(self)
        self._subscribers.add(sub)
        return sub

    def _unregister(self, sub: _Subscriber) -> None:
        self._subscribers.discard(sub)

    # ── Optional FastAPI integration ────────────────────────────────────────
    def fastapi_router(
        self,
        *,
        trigger_path: str = "/trigger",
        events_path: str = "/trigger-events",
    ) -> Any:
        """Return a FastAPI ``APIRouter`` exposing the standard trigger surface.

        Wire-up::

            from fastapi import FastAPI
            from saa_proactive import TriggerHub

            app = FastAPI()
            hub = TriggerHub()
            app.include_router(hub.fastapi_router())

        Requires ``attenlabs-saa-proactive[fastapi]``.
        """
        try:
            from fastapi import APIRouter, Body, HTTPException  # type: ignore
            from fastapi.responses import StreamingResponse  # type: ignore
        except ImportError as exc:  # pragma: no cover
            raise RuntimeError(
                "fastapi is required for TriggerHub.fastapi_router(). "
                "Install with: pip install attenlabs-saa-proactive[fastapi]"
            ) from exc

        router = APIRouter()

        @router.post(trigger_path)
        async def _trigger(payload: dict = Body(...)) -> dict:
            instructions = payload.get("instructions")
            if not isinstance(instructions, str) or not instructions.strip():
                raise HTTPException(400, "missing 'instructions' (string)")
            extra = {k: v for k, v in payload.items() if k != "instructions"}
            fanout = self.publish(instructions, **extra)
            if fanout == 0:
                # No subscriber connected to the events endpoint. The
                # operator's webhook should see a hard failure here —
                # silently swallowing the trigger would hide a real
                # misconfiguration (no browser open, SSE stream timed
                # out, etc.). Matches the posture of the reference
                # proactive-agent overlays.
                raise HTTPException(
                    503, f"no subscribers connected to {events_path}"
                )
            return {"ok": True, "subscribers": fanout}

        @router.get(events_path)
        async def _events() -> Any:
            sub = self.subscribe()
            return StreamingResponse(
                sub.sse_lines(), media_type="text/event-stream"
            )

        return router


__all__ = ["TriggerHub", "TriggerEvent"]
